File: src/frontend/src/services/anime.ts
import type { Anime, AnimeSeason, AnimeStatus } from "../types/anime";

const SHOWS_PAGE_SIZE = 50;

// The exact shape FastAPI sends, snake_case, matching the Python model
// field-for-field. Nothing outside this file should ever see raw backend
// data directly.
interface BackendSeason {
  id: string;
  show_id: string;
  season_number: number;
  name: string | null;
  episode_count: number | null;
  episodes_watched: number;
  status: string;
  air_date: string | null;
  poster_url: string | null;
  created_at: number;
  updated_at: number;
}

export interface BackendAnime {
  id: string;
  user_id: string;
  title: string;
  sort_title: string;
  description: string | null;
  first_air_date: string | null;
  episode_runtime_minutes: number | null;
  studios: string[];
  countries: string[];
  languages: string[];
  genres: string[];
  tags: string[];
  features: string[];
  age_rating: string | null;
  anilist_score: number | string | null;
  mal_score: number | string | null;
  source: string | null;
  poster_url: string | null;
  status: string;
  priority: string | null;
  favorite: boolean;
  rewatches: number;
  rating_story: number | string | null;
  rating_performance: number | string | null;
  rating_soundtrack: number | string | null;
  rating_overall: number | string | null;
  personal_rank: number | null;
  seasons: BackendSeason[];
  // unix timestamps in seconds, not ISO strings
  created_at: number;
  updated_at: number;
}

// Pydantic can serialize a Decimal as either a JSON number or a string
// depending on config, handle both rather than assume one
function toNumberOrNull(value: number | string | null): number | null {
  return value === null ? null : Number(value);
}

function unixSecondsToIso(seconds: number): string {
  return new Date(seconds * 1000).toISOString();
}

// backend sends "IN_PROGRESS", "WISHLIST", etc., frontend expects
// 'in progress', 'wishlist' (lowercase, spaces not underscores)
function normalizeStatus(raw: string): AnimeStatus {
  return raw.toLowerCase().replace(/_/g, " ") as AnimeStatus;
}
// inverse of normalizeStatus, 'in progress' -> 'IN_PROGRESS'
function denormalizeStatus(status: AnimeStatus): string {
  return status.toUpperCase().replace(/ /g, "_");
}

function mapBackendSeason(raw: BackendSeason): AnimeSeason {
  return {
    id: raw.id,
    showId: raw.show_id,
    seasonNumber: raw.season_number,
    name: raw.name,
    episodeCount: raw.episode_count,
    episodesWatched: raw.episodes_watched,
    status: normalizeStatus(raw.status),
    airDate: raw.air_date,
    posterUrl: raw.poster_url,
    createdAt: unixSecondsToIso(raw.created_at),
    updatedAt: unixSecondsToIso(raw.updated_at),
  };
}

export function mapBackendAnime(raw: BackendAnime): Anime {
  return {
    id: raw.id,
    userId: raw.user_id,
    title: raw.title,
    sortTitle: raw.sort_title,
    description: raw.description,
    firstAirDate: raw.first_air_date,
    episodeRuntimeMinutes: raw.episode_runtime_minutes,
    studios: raw.studios,
    countries: raw.countries,
    languages: raw.languages,
    genres: raw.genres,
    tags: raw.tags,
    features: raw.features,
    ageRating: raw.age_rating,
    anilistScore: toNumberOrNull(raw.anilist_score),
    malScore: toNumberOrNull(raw.mal_score),
    source: raw.source,
    posterUrl: raw.poster_url,
    status: normalizeStatus(raw.status),
    priority: raw.priority,
    favorite: raw.favorite,
    rewatches: raw.rewatches,
    ratingStory: toNumberOrNull(raw.rating_story),
    ratingPerformance: toNumberOrNull(raw.rating_performance),
    ratingSoundtrack: toNumberOrNull(raw.rating_soundtrack),
    ratingOverall: toNumberOrNull(raw.rating_overall),
    personalRank: raw.personal_rank,
    seasons: raw.seasons.map(mapBackendSeason),
    createdAt: unixSecondsToIso(raw.created_at),
    updatedAt: unixSecondsToIso(raw.updated_at),
  };
}

async function handle<T>(response: Response, action: string): Promise<T> {
  if (!response.ok) {
    const message = await response.text();
    throw new Error(`Failed to ${action}: ${response.status} ${message}`);
  }
  return response.json();
}

export async function fetchAnime(): Promise<Anime[]> {
  const all: BackendAnime[] = [];
  let skip = 0;
  while (true) {
    const response = await fetch(
      `/api/anime/list?skip=${skip}&limit=${SHOWS_PAGE_SIZE}`,
      { credentials: "include" },
    );
    const page = await handle<BackendAnime[]>(response, "fetch anime");
    all.push(...page);
    if (page.length < SHOWS_PAGE_SIZE) break;
    skip += SHOWS_PAGE_SIZE;
  }
  return all.map(mapBackendAnime);
}

export async function getAnime(id: string): Promise<Anime> {
  const response = await fetch(`/api/anime/get/${id}`, {
    credentials: "include",
  });
  const raw = await handle<BackendAnime>(response, `fetch anime ${id}`);
  return mapBackendAnime(raw);
}

export interface SeasonInput {
  seasonNumber: number;
  name?: string | null;
  episodeCount?: number | null;
  airDate?: string | null;
  posterUrl?: string | null;
}

function seasonInputToBody(input: SeasonInput): Record<string, unknown> {
  return {
    season_number: input.seasonNumber,
    name: input.name ?? null,
    episode_count: input.episodeCount ?? null,
    air_date: input.airDate ?? null,
    poster_url: input.posterUrl ?? null,
  };
}

export interface AnimeInput {
  title: string;
  description?: string | null;
  firstAirDate?: string | null;
  episodeRuntimeMinutes?: number | null;
  studios?: string[];
  countries?: string[];
  languages?: string[];
  genres?: string[];
  tags?: string[];
  features?: string[];
  ageRating?: string | null;
  anilistScore?: number | null;
  malScore?: number | null;
  source?: string | null;
  posterUrl?: string | null;
  status?: AnimeStatus;
  priority?: string | null;
  favorite?: boolean;
  rewatches?: number;
  ratingStory?: number | null;
  ratingPerformance?: number | null;
  ratingSoundtrack?: number | null;
  ratingOverall?: number | null;
  personalRank?: number | null;
  // omitted entirely (not just an empty array) means "auto-create a
  // default Season 1" — see create_anime on the backend
  seasons?: SeasonInput[];
}

// Round-trips a loaded Anime back into AnimeInput shape — used when a
// caller needs to change one field (e.g. toggling favorite from the
// detail page) without reopening the full edit form, since updateAnime
// always sends every field rather than a true partial patch.
export function animeToInput(show: Anime): AnimeInput {
  return {
    title: show.title,
    description: show.description,
    firstAirDate: show.firstAirDate,
    episodeRuntimeMinutes: show.episodeRuntimeMinutes,
    studios: show.studios,
    countries: show.countries,
    languages: show.languages,
    genres: show.genres,
    tags: show.tags,
    features: show.features,
    ageRating: show.ageRating,
    anilistScore: show.anilistScore,
    malScore: show.malScore,
    source: show.source,
    posterUrl: show.posterUrl,
    status: show.status,
    priority: show.priority,
    favorite: show.favorite,
    rewatches: show.rewatches,
    ratingStory: show.ratingStory,
    ratingPerformance: show.ratingPerformance,
    ratingSoundtrack: show.ratingSoundtrack,
    ratingOverall: show.ratingOverall,
    personalRank: show.personalRank,
  };
}

function inputToBody(input: AnimeInput): Record<string, unknown> {
  const body: Record<string, unknown> = {
    title: input.title,
    description: input.description ?? null,
    first_air_date: input.firstAirDate ?? null,
    episode_runtime_minutes: input.episodeRuntimeMinutes ?? null,
    studios: input.studios ?? [],
    countries: input.countries ?? [],
    languages: input.languages ?? [],
    genres: input.genres ?? [],
    tags: input.tags ?? [],
    features: input.features ?? [],
    age_rating: input.ageRating ?? null,
    anilist_score: input.anilistScore ?? null,
    mal_score: input.malScore ?? null,
    source: input.source ?? null,
    poster_url: input.posterUrl ?? null,
    priority: input.priority ?? null,
    favorite: input.favorite ?? false,
    rewatches: input.rewatches ?? 0,
    rating_story: input.ratingStory ?? null,
    rating_performance: input.ratingPerformance ?? null,
    rating_soundtrack: input.ratingSoundtrack ?? null,
    rating_overall: input.ratingOverall ?? null,
    personal_rank: input.personalRank ?? null,
  };
  if (input.status) body.status = denormalizeStatus(input.status);
  if (input.seasons) body.seasons = input.seasons.map(seasonInputToBody);
  return body;
}

export async function createAnime(input: AnimeInput): Promise<Anime> {
  const response = await fetch("/api/anime/create", {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(inputToBody(input)),
  });
  const raw = await handle<BackendAnime>(response, "create anime");
  return mapBackendAnime(raw);
}

export async function updateAnime(
  id: string,
  input: AnimeInput,
): Promise<Anime> {
  const response = await fetch(`/api/anime/update/${id}`, {
    method: "PATCH",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(inputToBody(input)),
  });
  const raw = await handle<BackendAnime>(response, `update anime ${id}`);
  return mapBackendAnime(raw);
}

export async function deleteAnime(id: string): Promise<void> {
  const response = await fetch(`/api/anime/delete/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!response.ok && response.status !== 204) {
    throw new Error(`Failed to delete anime ${id}: ${response.status}`);
  }
}

export async function createSeason(
  showId: string,
  input: SeasonInput,
): Promise<Anime> {
  const response = await fetch(`/api/anime/${showId}/seasons`, {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(seasonInputToBody(input)),
  });
  const raw = await handle<BackendAnime>(response, "create season");
  return mapBackendAnime(raw);
}

export interface SeasonUpdateInput {
  seasonNumber?: number;
  name?: string | null;
  episodeCount?: number | null;
  episodesWatched?: number;
  airDate?: string | null;
  posterUrl?: string | null;
  status?: AnimeStatus;
}

export async function updateSeason(
  showId: string,
  seasonId: string,
  input: SeasonUpdateInput,
): Promise<Anime> {
  const body: Record<string, unknown> = {};
  if ("seasonNumber" in input) body.season_number = input.seasonNumber;
  if ("name" in input) body.name = input.name;
  if ("episodeCount" in input) body.episode_count = input.episodeCount;
  if ("episodesWatched" in input) body.episodes_watched = input.episodesWatched;
  if ("airDate" in input) body.air_date = input.airDate;
  if ("posterUrl" in input) body.poster_url = input.posterUrl;
  if (input.status) body.status = denormalizeStatus(input.status);

  const response = await fetch(`/api/anime/${showId}/seasons/${seasonId}`, {
    method: "PATCH",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  const raw = await handle<BackendAnime>(response, "update season");
  return mapBackendAnime(raw);
}

export async function deleteSeason(
  showId: string,
  seasonId: string,
): Promise<Anime> {
  const response = await fetch(`/api/anime/${showId}/seasons/${seasonId}`, {
    method: "DELETE",
    credentials: "include",
  });
  const raw = await handle<BackendAnime>(response, "delete season");
  return mapBackendAnime(raw);
}

// A raw metadata search result, straight from whichever provider
// (AniList or MyAnimeList) found it — already snake_case-to-camelCase
// mapped here since these never round-trip back to the backend the way
// BackendAnime does. No `seasons` field: unlike TMDB for TV, neither
// AniList nor Jikan returns a season breakdown, so a new anime always
// gets the backend's auto-created default season instead.
export interface AnimeMetadataResult {
  provider: string;
  providerId: string;
  title: string;
  description: string | null;
  firstAirDate: string | null;
  episodeRuntimeMinutes: number | null;
  episodeCount: number | null;
  studios: string[];
  countries: string[];
  genres: string[];
  posterUrl: string | null;
  anilistScore: number | null;
  malScore: number | null;
  url: string | null;
}

export interface AnimeMetadataSearchResponse {
  query: string;
  providers: string[];
  providerErrors: string[];
  results: AnimeMetadataResult[];
}

interface BackendAnimeMetadataResult {
  provider: string;
  provider_id: string;
  title: string;
  description: string | null;
  first_air_date: string | null;
  episode_runtime_minutes: number | null;
  episode_count: number | null;
  studios: string[];
  countries: string[];
  genres: string[];
  poster_url: string | null;
  anilist_score: number | string | null;
  mal_score: number | string | null;
  url: string | null;
}

interface BackendAnimeMetadataSearchResponse {
  query: string;
  providers: string[];
  provider_errors: string[];
  results: BackendAnimeMetadataResult[];
}

export async function searchAnimeMetadata(
  query: string,
  limit = 8,
): Promise<AnimeMetadataSearchResponse> {
  const params = new URLSearchParams({ query, limit: String(limit) });
  const response = await fetch(`/api/anime/metadata/search?${params}`, {
    credentials: "include",
  });
  const raw = await handle<BackendAnimeMetadataSearchResponse>(
    response,
    "search anime metadata",
  );
  return {
    query: raw.query,
    providers: raw.providers,
    providerErrors: raw.provider_errors,
    results: raw.results.map((r) => ({
      provider: r.provider,
      providerId: r.provider_id,
      title: r.title,
      description: r.description,
      firstAirDate: r.first_air_date,
      episodeRuntimeMinutes: r.episode_runtime_minutes,
      episodeCount: r.episode_count,
      studios: r.studios,
      countries: r.countries,
      genres: r.genres,
      posterUrl: r.poster_url,
      anilistScore: toNumberOrNull(r.anilist_score),
      malScore: toNumberOrNull(r.mal_score),
      url: r.url,
    })),
  };
}

File: src/backend/src/database/models/anime.py
import time
from datetime import date
from decimal import Decimal
from enum import Enum
from typing import TYPE_CHECKING
from uuid import UUID, uuid4

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    Enum as SAEnum,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.database.base import Base

if TYPE_CHECKING:
    from src.database.models.user import User


class AnimeStatus(str, Enum):
    """Anime status aligned with a media library workflow — same value
    set as MovieStatus/TVShowStatus, shared here since a season also
    uses it."""

    DROPPED = "DROPPED"
    WISHLIST = "WISHLIST"
    WATCHLIST = "WATCHLIST"
    BACKLOG = "BACKLOG"
    IN_PROGRESS = "IN_PROGRESS"
    WATCHED = "WATCHED"
    FAVORITE = "FAVORITE"
    REWATCH = "REWATCH"


class Anime(Base):
    __tablename__ = "anime"

    # ------------------------------------------------------------------
    # Identity
    # ------------------------------------------------------------------

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)

    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user: Mapped["User"] = relationship()

    # set instead of actually deleting the row — same soft-delete
    # convention as Movie/TVShow/Game/Card. NULL means active/not deleted.
    deleted_at: Mapped[int | None] = mapped_column(BigInteger, nullable=True)

    # ------------------------------------------------------------------
    # Basic metadata
    # ------------------------------------------------------------------

    title: Mapped[str] = mapped_column(String(500), nullable=False)
    sort_title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    first_air_date: Mapped[date | None] = mapped_column(Date, nullable=True)

    # average minutes per episode
    episode_runtime_minutes: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # the animation studio(s) — anime's equivalent of TVShow.creators as
    # the primary credited-work field
    studios: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)

    countries: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)
    languages: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)
    genres: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)
    tags: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)
    features: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)

    age_rating: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # the media sub-format (TV, Movie, OVA, ONA, Special, Music) as
    # reported by AniList/Jikan — distinct from `kind`, which is fixed to
    # "anime" for every row in this table; this is what actually varies
    # per entry and drives the "TV"/"Movie"/"OVA" label shown in the UI
    format: Mapped[str | None] = mapped_column(String(20), nullable=True)

    # two external scores, matching how Movie/TVShow keep one provider's
    # score (tmdb_score) — here AniList and MyAnimeList (via Jikan) are
    # both real, independent sources worth keeping separately rather than
    # merging into one field
    anilist_score: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)
    mal_score: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)

    source: Mapped[str | None] = mapped_column(String(50), nullable=True)
    # the source provider's own id for this show (MyAnimeList's id via
    # Jikan) — kept so episode sync can hit that exact show again later
    # instead of re-searching by title and hoping for an exact match
    external_id: Mapped[str | None] = mapped_column(String(50), nullable=True)
    # AniList's own id, kept separately from external_id (which is MAL's)
    # — Jikan/MyAnimeList is unreliable (rate limits, occasional outages),
    # so episode sync tries it first when known but falls back to AniList's
    # streamingEpisodes data when Jikan is unavailable or external_id is
    # unknown. Two independent episode sources instead of one single point
    # of failure.
    anilist_id: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # NULL until the airing-check loop has actually looked (nothing to
    # filter on yet); True/False once AniList's own nextAiringEpisode
    # presence has been checked at least once. Lets the frequent
    # airing-check loop skip a show entirely once it's known to have
    # finished, instead of re-querying it every cycle forever.
    is_airing: Mapped[bool | None] = mapped_column(Boolean, nullable=True)

    # a direct external URL (AniList's CDN / Jikan's image field), stored
    # as-is — same convention as Movie.poster_url, never downloaded/resized
    poster_url: Mapped[str | None] = mapped_column(Text, nullable=True)

    # a wide-format background image (AniList's bannerImage), distinct
    # from the portrait poster_url above — used for the detail page hero
    backdrop_url: Mapped[str | None] = mapped_column(Text, nullable=True)

    # names of fields an admin has manually changed away from what a
    # metadata provider last set — "Apply metadata" skips these instead
    # of silently overwriting a deliberate fix
    locked_fields: Mapped[list[str]] = mapped_column(ARRAY(String), nullable=False, default=list)

    # The Related tab's {chain, branches, recommendations} payload from
    # AniList, cached after the first fetch instead of re-fetched on every
    # visit — a single load can already mean a dozen+ AniList requests
    # (one per prequel/sequel chain hop, one per branch-group reorder
    # lookup), and AniList's rate limit is shared and low. NULL means
    # never fetched yet.
    relations_cache: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    relations_cached_at: Mapped[int | None] = mapped_column(BigInteger, nullable=True)

    # ------------------------------------------------------------------
    # Personal library state
    # ------------------------------------------------------------------

    status: Mapped[AnimeStatus] = mapped_column(
        SAEnum(AnimeStatus, native_enum=False, length=30),
        nullable=False,
        default=AnimeStatus.WISHLIST,
    )
    priority: Mapped[str | None] = mapped_column(String(20), nullable=True)
    favorite: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    rewatches: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)

    # ------------------------------------------------------------------
    # Ratings
    # ------------------------------------------------------------------

    rating_story: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)
    rating_performance: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)
    rating_soundtrack: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)
    rating_overall: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)

    personal_rank: Mapped[int | None] = mapped_column(Integer, nullable=True)

    # eager (selectin, not the default lazy) — AnimeRead serializes
    # `seasons` straight into the API response (same reason as
    # Game.links / TVShow.seasons: a lazy load can't run once the async
    # session that produced the row is out of the request's await context)
    seasons: Mapped[list["AnimeSeason"]] = relationship(
        back_populates="show",
        order_by="AnimeSeason.season_number",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    # ------------------------------------------------------------------
    # Timestamps
    # ------------------------------------------------------------------

    created_at: Mapped[int] = mapped_column(
        BigInteger, nullable=False, default=lambda: int(time.time())
    )
    updated_at: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        default=lambda: int(time.time()),
        onupdate=lambda: int(time.time()),
    )


class AnimeSeason(Base):
    """One season/cour of an Anime — a real child row, not just a count,
    so progress and status can be tracked per season independently of the
    show overall. Most anime will carry exactly one season row (a new
    cour is usually its own separate AniList entry rather than a season
    of an existing one). `episodes_watched`/`episode_count` stay as the
    flat progress numbers used everywhere else in the app; `episodes`
    below is the optional richer per-episode breakdown."""

    __tablename__ = "anime_seasons"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    show_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("anime.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    show: Mapped["Anime"] = relationship(back_populates="seasons")

    season_number: Mapped[int] = mapped_column(Integer, nullable=False)
    name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    episode_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    episodes_watched: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    status: Mapped[AnimeStatus] = mapped_column(
        SAEnum(AnimeStatus, native_enum=False, length=30, name="animeseasonstatus"),
        nullable=False,
        default=AnimeStatus.WISHLIST,
    )
    air_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    poster_url: Mapped[str | None] = mapped_column(Text, nullable=True)

    episodes: Mapped[list["AnimeEpisode"]] = relationship(
        back_populates="season",
        order_by="AnimeEpisode.episode_number",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    created_at: Mapped[int] = mapped_column(
        BigInteger, nullable=False, default=lambda: int(time.time())
    )
    updated_at: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        default=lambda: int(time.time()),
        onupdate=lambda: int(time.time()),
    )


class AnimeEpisode(Base):
    """One episode of an AnimeSeason. Rows are synced in from the source
    provider (Jikan/MyAnimeList) the first time a season's episode list
    is requested, then persisted here — later requests read straight from
    this table instead of re-fetching, and `watched`/`rating` are purely
    local, never touched by a re-sync."""

    __tablename__ = "anime_episodes"

    id: Mapped[UUID] = mapped_column(PG_UUID(as_uuid=True), primary_key=True, default=uuid4)
    season_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("anime_seasons.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    season: Mapped["AnimeSeason"] = relationship(back_populates="episodes")

    episode_number: Mapped[int] = mapped_column(Integer, nullable=False)
    title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    air_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    runtime_minutes: Mapped[int | None] = mapped_column(Integer, nullable=True)
    still_url: Mapped[str | None] = mapped_column(Text, nullable=True)

    watched: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    rating: Mapped[Decimal | None] = mapped_column(Numeric(4, 2), nullable=True)

    created_at: Mapped[int] = mapped_column(
        BigInteger, nullable=False, default=lambda: int(time.time())
    )
    updated_at: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        default=lambda: int(time.time()),
        onupdate=lambda: int(time.time()),
    )
